Sources/Machine/py_abstract_machine.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright 2019 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Machine/py_abstract_machine.hpp"
+
+#include <cstdio>
+
+#include <pybind11/complex.h>
+#include <pybind11/eigen.h>
+#include <pybind11/pybind11.h>
+
+#include "Utils/messages.hpp"
+
+namespace netket {
+
+bool ShouldIDoIO() noexcept {
+  auto rank = 0;
+  auto const status = MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  if (status == MPI_SUCCESS) {
+    return rank == 0;
+  }
+  std::fprintf(stderr,
+               "[NetKet] MPI_Comm_rank failed: doing I/O on all processes.\n");
+  return true;
+}
+
+template <class Function, class... Args>
+auto ShouldNotThrow(Function &&function, Args &&... args) noexcept
```

Comment:
What is this?!?!

Reply:
It's a wrapper function which handles exceptions.

Reply:
Maybe this should be moved somewhere else?

Reply:
No, because it is specific for that file. The rest of the code base doesn't call Python and algorithms should just be implemented properly if they're marked `noexcept`.

I can move it to `detail` namespace.